src/lowering/desugar.ml

```diff
@@ -239,6 +242,53 @@ and lexp' = function
   | S.IdxE (e1, e2) -> I.IdxLE (exp e1, exp e2)
   | _ -> raise (Invalid_argument ("Unexpected expression as lvalue"))
 
+and rewrite_for_to_while p arr proj c0 c1 c2 e1 e2 =
+  (* for (p in e1) e2 when e1 = ⟨...array-typed...⟩.vals()
+     ~~>
+     let arr = ⟨...array-typed...⟩ ;
+     let size = arr.size() ;
+     var indx = 0 ;
+     label l loop {
+       if indx < size
+       then { let p = arr[indx]; e2; indx += 1 }
+       else { break l }
+     } *)
+  let arr_typ = arr.note.note_typ in
+  let triv t = { note_typ = t; note_eff = T.Triv } in
+  let body arrb =
+    let size_exp =
+      exp { e1 with note = triv T.nat;
+                    it = CallE ({ c0 with note = triv T.(Func (Local, Returns, [], [], [nat]));
+                                          it = DotE (arrb, { proj with it = "size" }) },
+                                c1, c2) } in
+    let indx = fresh_var "indx" T.(Mut nat) in
+    let indexing_exp = exp
+           { note = triv T.(as_immut (as_array arr_typ));
+             at = arr.at;
+             it = IdxE (arrb, { arr with note = triv T.nat;
+                                         it = VarE { it = id_of_var indx;
+                                                     note = ();
+                                                     at = arr.at } }) } in
+    let size = fresh_var "size" T.nat in
+    blockE [ letD size size_exp
+           ; varD indx (natE Numerics.Nat.zero)]
+      (whileE (primE (Ir_def.Ir.RelPrim (T.nat, LtOp))
+                 [varE indx; varE size])
+         (blockE [ letP p indexing_exp
+                 ; expD e2]
+            (assignE indx
+               (primE (Ir_def.Ir.BinPrim (T.nat, AddOp))
+                  [ varE indx
+                  ; natE (Numerics.Nat.of_int 1)])))) in
+  let arr_ir = exp arr in
+  match arr_ir.it with
+  | I.VarE v when p.it <> I.VarP v -> body arr
+  | _ ->
+    let arrv = fresh_var "arr" arr_typ in
+    letE arrv arr_ir
+      (body { arr with it = VarE { it = id_of_var arrv; note = (); at = arr.at };
+                       note = triv arr.note.note_typ })
```

Comment:
```suggestion
      (body (varE arrv))
```
or maybe better to abstract `body` on a `var`, not `exp`, so you know the effect is trivial (this is call-by-value) and call:

```
body arrv
````